code/modules/admin/ipintel.dm

```diff
@@ -0,0 +1,224 @@
+	var/cache = FALSE
+	var/cacheminutesago = 0
+	var/cachedate = ""
+	var/cacherealtime = 0
+
+/datum/ipintel/New()
+	cachedate = SQLtime()
+	cacherealtime = world.realtime
+
+/datum/ipintel/proc/is_valid()
+	. = FALSE
+	if(intel < 0)
+		return
+	if(intel <= config.ipintel_rating_bad)
+		if(world.realtime < cacherealtime + (config.ipintel_save_good * 60 * 60 * 10))
+			return TRUE
+	else
+		if(world.realtime < cacherealtime + (config.ipintel_save_bad * 60 * 60 * 10))
+			return TRUE
+
+/proc/get_ip_intel(ip, bypasscache = FALSE, updatecache = TRUE)
+	var/datum/ipintel/res = new()
+	res.ip = ip
+	. = res
+	if(!ip || !config.ipintel_email || !SSipintel.enabled)
+		return
+	if(!bypasscache)
+		var/datum/ipintel/cachedintel = SSipintel.cache[ip]
+		if(cachedintel && cachedintel.is_valid())
+			cachedintel.cache = TRUE
+			return cachedintel
+
+		if(dbcon.IsConnected())
+			var/rating_bad = config.ipintel_rating_bad
+			var/DBQuery/query_get_ip_intel = dbcon.NewQuery({"
+				SELECT date, intel, TIMESTAMPDIFF(MINUTE,date,NOW())
+				FROM [format_table_name("ipintel")]
+				WHERE
+					ip = INET_ATON('[ip]')
+					AND ((
+							intel < [rating_bad]
+							AND
+							date + INTERVAL [config.ipintel_save_good] HOUR > NOW()
+						) OR (
+							intel >= [rating_bad]
+							AND
+							date + INTERVAL [config.ipintel_save_bad] HOUR > NOW()
+					))
+				"})
+			if(!query_get_ip_intel.Execute())
+				qdel(query_get_ip_intel)
+				return
+			if(query_get_ip_intel.NextRow())
+				res.cache = TRUE
+				res.cachedate = query_get_ip_intel.item[1]
+				res.intel = text2num(query_get_ip_intel.item[2])
+				res.cacheminutesago = text2num(query_get_ip_intel.item[3])
+				res.cacherealtime = world.realtime - (text2num(query_get_ip_intel.item[3])*10*60)
+				SSipintel.cache[ip] = res
+				qdel(query_get_ip_intel)
+				return
+			qdel(query_get_ip_intel)
+	res.intel = ip_intel_query(ip)
+	if(updatecache && res.intel >= 0)
+		SSipintel.cache[ip] = res
+		if(dbcon.IsConnected())
+			var/DBQuery/query_add_ip_intel = dbcon.NewQuery("INSERT INTO [format_table_name("ipintel")] (ip, intel) VALUES (INET_ATON('[ip]'), [res.intel]) ON DUPLICATE KEY UPDATE intel = VALUES(intel), date = NOW()")
+			query_add_ip_intel.Execute()
+			qdel(query_add_ip_intel)
+
+
+/proc/ip_intel_query(ip, var/retryed=0)
+	. = -1 //default
+	if(!ip)
+		return
+	if(SSipintel.throttle > world.timeofday)
+		return
+	if(!SSipintel.enabled)
+		return
+
+	var/list/http[] = world.Export("http://[config.ipintel_domain]/check.php?ip=[ip]&contact=[config.ipintel_email]&format=json&flags=b")
+
+	if(http)
+		var/status = text2num(http["STATUS"])
+
+		if(status == 200)
+			var/response = json_decode(file2text(http["CONTENT"]))
+			if(response)
+				if(response["status"] == "success")
+					var/intelnum = text2num(response["result"])
+					if(isnum(intelnum))
+						return text2num(response["result"])
+					else
+						ipintel_handle_error("Bad intel from server: [response["result"]].", ip, retryed)
```

Comment:
Reviewer's note:
I'm glad that this case is here, in event that whatever server it's talking to turns out to be a phony via DNS shenanigans or something